### PR TITLE
Unsend Requests

### PR DESCRIPTION
**Pull request by RyanRory**
- Use the `delete` endpoint to delete a specific message with the hash
- Send `UnsendRequest` to the recipient and linked devices to delete the message
- Dismiss the notification of a deleted message
- New action sheet UI for delete options
- New message bubble UI for deleted messages

### CODE REVIEW

File: Session/Conversations/ConversationVC+Interaction.swift [hunk 1]
```diff
@@ -555,7 +555,81 @@ extension ConversationVC : InputViewDelegate, MessageCellDelegate, ContextMenuAc
...
+        case .outgoingMessage: unsendRequest.author = getUserHexEncodedPublicKey()
+        default: return nil// Should never occur
+        }
+        unsendRequest.timestamp = viewItem.interaction.timestamp
+        return unsendRequest
+    }
+    
+    func deleteLocally(_ viewItem: ConversationViewItem) {
+        viewItem.deleteLocallyAction()
+        if let unsendRequest = buildUsendRequest(viewItem) {
```
Comment (nielsandriesse) [inline Session/Conversations/ConversationVC+Interaction.swift:618]: Why does delete locally send an unsend request?

File: Session/Conversations/ConversationViewModel.m [hunk 2]
```diff
@@ -830,6 +830,17 @@ - (void)updateViewWithOldItemIdList:(NSArray<NSString *> *)oldItemIdList
                 [updatedNeighborItemSet addObject:itemId];
             }
         }
+        
+        // Add the following item of a deleted message to update
+        // to show the date header of the deleted message if needed
+        for (NSString *deletedItemId in deletedItemIdSet) {
```
Comment (nielsandriesse) [inline Session/Conversations/ConversationViewModel.m:836]: Is this still needed even with the recent message updating fix?

File: SessionMessagingKit/Sending & Receiving/MessageReceiver+Handling.swift [hunk 3]
```diff
@@ -218,6 +220,30 @@ extension MessageReceiver {
...
     
     
+    // MARK: - Unsend Requests
+    
+    public static func handleUnsendRequest(_ message: UnsendRequest, using transaction: Any) {
+        guard message.sender == message.author else { return }
+        let userPublicKey = getUserHexEncodedPublicKey()
+        let transaction = transaction as! YapDatabaseReadWriteTransaction
+        if let author = message.author, let timestamp = message.timestamp,
+           let messageToDelete = userPublicKey == message.sender ? TSOutgoingMessage.find(withTimestamp: timestamp) : TSIncomingMessage.find(withAuthorId: author, timestamp: timestamp, transaction: transaction) {
```
Comment (nielsandriesse) [inline SessionMessagingKit/Sending & Receiving/MessageReceiver+Handling.swift:230]: Would be nice to split this line up a bit

Comment (nielsandriesse) [reply SessionMessagingKit/Sending & Receiving/MessageReceiver+Handling.swift:230]:
```
let messageToDelete: TSMessage
if ... {
    message = TSOutgoingMessage...
} else {
    message = TSIncomingMessage...
}
```

File: Session/Conversations/ConversationVC+Interaction.swift (hunk 1 above)
Comment (RyanRory) [reply Session/Conversations/ConversationVC+Interaction.swift:618]: For linked device to delete this message

Comment (RyanRory) [reply Session/Conversations/ConversationVC+Interaction.swift:618]: > Why does delete locally send an unsend request?

For linked device to delete this message

File: Session/Conversations/ConversationViewModel.m (hunk 2 above)
Comment (RyanRory) [reply Session/Conversations/ConversationViewModel.m:836]: > Is this still needed even with the recent message updating fix?

Yeah, I think so. With the MessageInvalidator, the updated items won't include the neighbours but only the deleted message itself.

File: Session/Conversations/ConversationVC.swift [hunk 4]
```diff
@@ -4,7 +4,8 @@
 // • Photo rounding (the small corners don't have the correct rounding)
 // • Remaining search glitchiness
 
-final class ConversationVC : BaseVC, ConversationViewModelDelegate, OWSConversationSettingsViewDelegate, ConversationSearchControllerDelegate, UITableViewDataSource, UITableViewDelegate {
+final class ConversationVC : BaseVC, ConversationViewModelDelegate, OWSConversationSettingsViewDelegate, ConversationSearchControllerDelegate, UITableViewDataSource, UITableViewDelegate {    
+    let isUnsendRequesEnabled = false // Switch this to true if unsend request is done on all platforms
```
Comment (malmazuke) [inline Session/Conversations/ConversationVC.swift:8]: Minor typo here

File: Session/Conversations/ConversationVC+Interaction.swift [hunk 5]
```diff
@@ -555,7 +555,83 @@ extension ConversationVC : InputViewDelegate, MessageCellDelegate, ContextMenuAc
...
+            
+            self.inputAccessoryView?.isHidden = true
+            self.inputAccessoryView?.alpha = 0
+            self.presentAlert(alertVC)
+        } else {
+            deleteLocally(viewItem)
+        }
+    }
+    
+    private func buildUsendRequest(_ viewItem: ConversationViewItem) -> UnsendRequest? {
```
Comment (malmazuke) [inline Session/Conversations/ConversationVC+Interaction.swift:602]: Minor typo

File: Session/Conversations/ConversationVC.swift (hunk 4 above)
Comment (nielsandriesse) [reply Session/Conversations/ConversationVC.swift:8]: Spotted that and the below typo too haha. Have fixes for them ready :)

**Comment by nielsandriesse** [comment]
Will fix a few of the typos on dev

**Pull request merged**